### PR TITLE
[codex] Add RIR verifier MVP

### PR DESCRIPTION
**Pull request by hurricane1026**
## Summary

Adds a small native verifier entry point for RIR route graphs. The verifier checks structural invariants that must hold before Rut can reason about stronger runtime properties.

Covered in this MVP:

- final terminator shape for every block
- branch and jump target validity
- default reachability from entry
- yield terminator count matching function metadata
- yield kind range validation against the runtime ABI
- basic state-zero and explicit resume block bounds

The formal verification notes now describe this as the current native verifier foundation, separate from broader TLA+/deadlock/progress modeling.

## Validation

- `cmake --build build --target test_rir`
- `./build/tests/test_rir`
- `git diff --check`
- `git diff --cached --check`

### CODE REVIEW

File: include/rut/compiler/verifier.h [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        if (term.op == Opcode::Br) {
+            const u32 targets[2] = {term.imm.block_targets[0].id, term.imm.block_targets[1].id};
+            for (u32 i = 0; i < 2; i++) {
+                summary.branch_edge_count++;
+                if (!reachable[targets[i]]) {
+                    reachable[targets[i]] = true;
+                    worklist[work_end++] = targets[i];
+                }
+            }
+        } else if (term.op == Opcode::Jmp) {
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:330]: Fixed in `da7c605`: `YieldTimer` continuations now participate in verifier reachability. For explicit resume tables, the verifier treats the encoded `next_state` as an edge to `resume_blocks[next_state]`, so valid wait/resume routes no longer get rejected as unreachable.

File: include/rut/compiler/verifier.h [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            if (!verify_valid_block_target(*fn, term.imm.block_targets[0])) {
+                return verify_fail(summary,
+                                   VerifyIssueCode::InvalidJumpTarget,
+                                   function_index,
+                                   bi,
+                                   block.inst_count - 1,
+                                   term.imm.block_targets[0].id);
+            }
+        } else if (term.is_yield()) {
+            seen_yield_terminators++;
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:250]: Fixed in `da7c605`: the verifier now decodes `YieldTimer`'s packed `yield_kind`, `next_state`, and payload from `imm.i64_val`, validates kind/state bounds, and checks the encoded fields against the function yield metadata.

File: include/rut/compiler/verifier.h [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            if (!verify_valid_block_target(*fn, term.imm.block_targets[0])) {
+                return verify_fail(summary,
+                                   VerifyIssueCode::InvalidJumpTarget,
+                                   function_index,
+                                   bi,
+                                   block.inst_count - 1,
+                                   term.imm.block_targets[0].id);
+            }
+        } else if (term.is_yield()) {
+            if (term.op == Opcode::YieldTimer) {
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:210]: Fixed in `5be2531`: the verifier now rejects non-lowered yield terminators (`YieldHttpGet`, `YieldHttpPost`, `YieldForward`) with `UnsupportedYieldTerminator`. This keeps the verifier scoped to the RIR shape that current JIT codegen can execute, while `YieldTimer` remains the validated lowered state-machine boundary.

**Comment by hurricane1026** [comment]
@codex，review

File: include/rut/compiler/verifier.h [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+                if (next_state == 0 || next_state > fn->yield_count) {
+                    return verify_fail(summary,
+                                       VerifyIssueCode::InvalidYieldNextState,
+                                       function_index,
+                                       bi,
+                                       block.inst_count - 1,
+                                       next_state);
+                }
+            }
+            seen_yield_terminators++;
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:250]: Fixed in dabc772: the verifier now tracks per-yield-state coverage while scanning YieldTimer terminators. It rejects duplicate encoded next_state values with DuplicateYieldNextState, keeps an explicit MissingYieldNextState guard, and caps yield states to the RIR resume-table bound so the coverage proof stays complete. I also added a focused RIR unit test plus frontend e2e coverage that lowers a two-wait route and mutates the second yield to duplicate state 1.

File: include/rut/compiler/verifier.h [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+            }
+
+            if (fn->has_explicit_resume_blocks) {
+                const u32 target = fn->resume_blocks[next_state];
+                summary.branch_edge_count++;
+                if (!reachable[target]) {
+                    reachable[target] = true;
+                    worklist[work_end++] = target;
+                }
+            } else if (fn->state_zero_enters_entry) {
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:356]: Fixed in ca1217e: verifier now rejects yield state machines that have YieldTimer terminators but no supported resume mapping (`has_explicit_resume_blocks == false` and `state_zero_enters_entry == false`) with `MissingYieldResumeMapping`. Added a focused RIR unit test for the malformed single-yield shape and frontend e2e coverage that lowers a wait route, clears the resume scheme, and checks the diagnostic text.

File: include/rut/compiler/verifier.h [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+            return verify_fail(summary, VerifyIssueCode::InvalidStateZeroEntry, function_index);
+        }
+    }
+    if (fn->has_explicit_resume_blocks) {
+        if (fn->yield_count >= Function::kMaxResumeBlocks) {
+            return verify_fail(summary, VerifyIssueCode::InvalidResumeBlock, function_index);
+        }
+        for (u32 i = 0; i <= fn->yield_count; i++) {
+            if (fn->resume_blocks[i] >= fn->block_count) {
+                return verify_fail(summary,
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:297]: Fixed in 897cd6d: in explicit-resume mode the verifier now requires `resume_blocks[0]` to match the function entry block. This keeps verifier reachability rooted at the same state-0 target used by JIT dispatch and rejects stale tables that would skip block 0 on initial requests. Added a focused RIR unit test and frontend e2e coverage that lowers a wait route, mutates `resume_blocks[0]`, and checks the `InvalidStateZeroEntry` diagnostic.

File: include/rut/compiler/verifier.h [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+    }
+    summary.yield_count = fn->yield_count;
+
+    if (fn->yield_count > 0 && !fn->has_explicit_resume_blocks && !fn->state_zero_enters_entry) {
+        return verify_fail(summary, VerifyIssueCode::MissingYieldResumeMapping, function_index);
+    }
+    if (fn->state_zero_enters_entry) {
+        if (fn->state_zero_entry_block >= fn->block_count ||
+            fn->resume_terminal_block >= fn->block_count) {
+            return verify_fail(summary, VerifyIssueCode::InvalidStateZeroEntry, function_index);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h:288]: Fixed in 055c13c: for non-explicit `state_zero_enters_entry` mode, the verifier now requires `state_zero_entry_block` to match the function entry block. That keeps verifier reachability rooted at the same state-0 target used by JIT dispatch. Added a focused RIR unit test plus frontend e2e coverage that lowers a wait route, mutates it into non-explicit state-zero mode with a stale entry target, and checks the `InvalidStateZeroEntry` diagnostic.

File: include/rut/compiler/verifier.h [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+    result.issue.code = code;
+    result.issue.function_index = function_index;
+    result.issue.block_index = block_index;
+    result.issue.inst_index = inst_index;
+    result.issue.target_index = target_index;
+    return result;
+}
+
+inline bool verify_valid_yield_kind(u8 kind) {
+    return kind <= static_cast<u8>(jit::YieldKind::UpstreamSend);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h]: Fixed in 616a20d: `verify_valid_yield_kind` is now an explicit allow-list of runtime-schedulable lowered wait/event kinds: `Timer`, `Any`, `Recv`, `Send`, `UpstreamConnect`, `UpstreamRecv`, and `UpstreamSend`. It rejects `HttpGet`, `HttpPost`, and `Forward` even though they are ABI enum values, matching `jit_dispatch` behavior. Added a RIR unit test and frontend e2e mutation test where encoded kind and metadata both use `HttpPost` and verifier reports `InvalidYieldKind`.

**Comment by hurricane1026** [comment]
@codex, review

**Comment by hurricane1026** [comment]
@codex review

File: include/rut/compiler/verifier.h [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+                                   fn->resume_blocks[i]);
+            }
+        }
+        if (fn->resume_blocks[0] != fn->blocks[0].id.id) {
+            return verify_fail(summary,
+                               VerifyIssueCode::InvalidStateZeroEntry,
+                               function_index,
+                               0,
+                               0,
+                               fn->resume_blocks[0]);
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h]: Fixed in 46e294d: the verifier no longer assumes reachability starts at blocks[0]. It now seeds reachability from the same state-0 dispatch target used by codegen: resume_blocks[0] in explicit-resume mode, state_zero_entry_block in non-explicit state-zero mode, otherwise blocks[0]. This allows a valid prelude state-zero target when it reaches the original entry through normal control flow, while stale targets that skip entry are still rejected as UnreachableBlock. Added a focused RIR prelude acceptance test and updated the stale-target RIR/frontend e2e tests. Verified with ./build/tests/test_rir, ./build/tests/test_frontend --filter=frontend.rir_verifier_e2e, and git diff --check.

**Comment by hurricane1026** [comment]
@codex review

File: include/rut/compiler/verifier.h [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+            if (next_state == 0 || next_state > fn->yield_count) {
+                return verify_fail(summary,
+                                   VerifyIssueCode::InvalidYieldNextState,
+                                   function_index,
+                                   bi,
+                                   block.inst_count - 1,
+                                   next_state);
+            }
+            const u32 yi = static_cast<u32>(next_state - 1);
+            if (fn->yield_kinds[yi] != kind || fn->yield_payload[yi] != payload) {
```
[the remark this answers is not in the document]
Comment (hurricane1026) [reply include/rut/compiler/verifier.h]: Fixed in 9b3a8e3: YieldTimer metadata validation now runs during the initial full-block verifier scan, before reachability traversal. The verifier first checks that yield metadata arrays exist, then compares each YieldTimer encoded kind/payload against the indexed metadata regardless of whether that block is reachable under the current VerifyOptions. The worklist pass now only adds continuation edges. I kept the metadata kind validity check after the structural scan so non-lowered yield terminators still report UnsupportedYieldTerminator instead of being shadowed by metadata prevalidation. Added a regression test with require_all_blocks_reachable=false where a dead yield block has mismatched payload metadata and is rejected with YieldMetadataMismatch. Verified with cmake --build build --target test_frontend test_rir, ./build/tests/test_rir, ./build/tests/test_frontend --filter=frontend.rir_verifier_e2e, and git diff --check.

**Comment by hurricane1026** [comment]
@codex review
